library/Core/src/Juvix/Core/Common/Context.hs

```diff
@@ -334,68 +359,100 @@ instance MapSym PrivNameSpace where
+  m (Maybe (T term ty sumRep))
+modifySpaceImp f symbol@(s :| ymbol) t =
+  -- check the current Module first, to properly determine which one to go down
+  case NameSpace.lookupInternal s (t ^. _currentNameSpace . contents) of
     Just (NameSpace.Pub _) ->
-      updateCurr t <$> recurse f (s :| ymbol) currentNameSpace
+      applyAndSetCurrent (recurseImp f (s :| ymbol))
     Just (NameSpace.Priv _) ->
-      updateCurr t . unPriv <$> recurse f (s :| ymbol) (Priv currentNameSpace)
+      applyAndSetCurrent (fmap (fmap unPriv) . recurseImp f symbol . Priv)
+    -- The first part of the name is not in the map, maybe they fully qualified the name
     Nothing ->
-      case NameSymbol.takePrefixOf currentName (removeTopName (s :| ymbol)) of
-        Just subPath ->
-          updateCurr t <$> recurse f subPath currentNameSpace
+      case NameSymbol.takePrefixOf (t ^. _currentName) (removeTopName symbol) of
+        -- currentName : Foo.Bar, symbol: TopLevel.Foo.Bar.baz
+        -- currentName : Foo.Bar, symbol: Foo.Bar.baz
+        -- both of these will hit this path
+        Just subPath -> do
+          applyAndSetCurrent (recurseImp f subPath)
+        -- For this path, we have
+        -- currentName : Foo.Bar, symbol: Something.Else
         Nothing ->
           -- we do one more match as we have to make sure
           -- if we are adding foo we add it to the right place
           case ymbol of
-            [] ->
-              updateCurr t <$> recurse f (s :| ymbol) currentNameSpace
+            -- This case is
+            -- currentName: Foo.Bar, symbol: s
+            -- so just apply on the current, we'll stop right away
+            [] -> applyAndSetCurrent (recurseImp f symbol)
+            -- in this case, we have more than just a single symbol
             (newS : newYmbol) ->
+              -- currentName: Foo.Bar, symbol: TopLevel.First.Rest
+              -- update the TopLevel on First.Rest!
               if  | s == topLevelName ->
-                    updateTopLevel t <$> recurse f (newS :| newYmbol) topLevelMap
-                  | otherwise ->
-                    updateTopLevel t <$> recurse f (s :| ymbol) topLevelMap
+                    applyAndSetTop (recurseImp f (newS :| newYmbol))
+                  -- currentName: Foo.Bar, symbol: First.MoreStuff
+                  -- just recurse on the symbol entirely
+                  | otherwise -> applyAndSetTop (recurseImp f symbol)
   where
-    updateCurr t newCurrent =
-      t {currentNameSpace = newCurrent}
-    updateTopLevel t newTop =
-      t {topLevelMap = newTop}
-
-recurse ::
-  MapSym m =>
+    -- dumb repeat code but idk how to handle to remove the tuple ☹
+    applyAndSetTop f =
+      t |> applyAndSet f (_topLevelMap, _topLevelMap)
+    applyAndSetCurrent f =
+      t |> applyAndSet f (_currentNameSpace . contents, _currentNameSpace . contents)
+
+applyAndSet f (l1, l2) t = do
+  ret <- f (t ^. l1)
+  pure ((\r -> set l2 r t) <$> ret)
+
```

Comment:
Can someone look at this function and tell me if you can understand the function with the comments?

That alone should be enough but I would like some opinions

Reply:
`modifySpaceImp` is very long. I think it'll be easier to read if you could split it into smaller functions. Also, could you give a bit more context in the comments?

Reply:
I only see one place to split it up, the Nothing case  (without comments the function is only 15 lines long!),, as I just do some map lookups.

But I do agree, I think the main thing that needs to be commented is `recurseImp`, since we just dispatch onto that.

Also I should comment on that we just update the map given to `recurseImp`.

However, the important part is, do the comments properly give you a feeling of what paths things take given what the current name is and the symbol name

(that is what the `currentName : Foo.Bar, symbol: Something.Else` tries to show)